### ISSUE TITLE
Update Rendering Mechanism

### ISSUE THREAD

**Issue by nikitaeverywhere** [intersystems-ru/LightPivotTable#12]
LPT's rendering mechanism must be updated to move further. These things need to be done:

- [ ] Fix rows/cols overriding each others style.
- [ ] Teach LPT to handle this-like formatting `#,#.##;-#,#.##;;;`, references #11.
- [ ] Issue with empty rows (to be clarified by @ohoferek).
- [ ] When there are multiple dimensions display left upper corner as blank or exactly as DSW displays. References #13.
- [ ] Discover timeout problem (to be clarified by @ohoferek).

**Comment by ondrejhoferek** [comment]
I have a note regarding the first thing on the list. The issue is that when different style is configured for the header and the row, such as in this case:

![formatting dsui](https://user-images.githubusercontent.com/16596359/34992101-4f21ebb6-facc-11e7-8d5d-611f03da711c.png)

LPT forces the cell style for the header as well:

![formatting dsw](https://user-images.githubusercontent.com/16596359/34992111-5577ca26-facc-11e7-8a16-2b47d18002b8.png)

**Comment by nikitaeverywhere** [comment]
Working on test pivot:

http://146.185.143.59:57772/csp/samples/_DeepSee.UI.Analyzer.zen?$NAMESPACE=SAMPLES&PIVOT=LPT%2FFormattingTest.pivot

![2018-01-23_121305](https://user-images.githubusercontent.com/4989256/35270112-ccb22c5c-0036-11e8-88ee-4088b72890ea.png)

VS

![2018-01-23_121343](https://user-images.githubusercontent.com/4989256/35270136-e11bc2e8-0036-11e8-8075-d45713bd0ea5.png)

The formatting works well on cells but not on headers.

**Comment by nikitaeverywhere** [comment]
@ohoferek, a problem with rows/cols formatting is kind of tough problem I faced a long time ago, didn't solve it completely and temporary made a useable "workaround" for half of the use cases. This issue may last for a while until we figure out how to deal with it.

I am pretty sure this can't work as expected without modifying MDX2JSON. I urge @eduard93 about MDX2JSON, please help :)

**The problem**: I am unable to merge `/mdx` endpoint results with `/DataSource` endpoint results to render the correct view **in terms of rows/cols formatting**: there's insufficient data to do it.

### Here's detailed explanation:

Look at [this test pivot](http://146.185.143.59/csp/samples/_DeepSee.UI.Analyzer.zen?$NAMESPACE=SAMPLES&PIVOT=LPT%2FFormattingTest.pivot). The goal is to make LPT to pick the headers/rows/cols formatting correctly, through it seems like this is not possible yet. Notice these row headers:

![2018-01-28_213235](https://user-images.githubusercontent.com/4989256/35486175-fd925b40-0472-11e8-9bf0-66b124265e15.png)

I've made them different colors to demonstrate the problem. Now, open this pivot in [DSW](http://146.185.143.59/dsw/#!/d/LPT/FormattingTest.dashboard) while Developer Tools opened. Notice the last two requests to `/DataSource` and `/mdx` endpoints.

The task is to take the `/mdx` result and style it using the data from `/DataSource` (which returns a pivot definition). In other words, merge results from `/mdx` and `/DataSource` into a single data structure. And the problem is to find the correspondence of how to apply the right definition to the right (tuple, header) dimension and its children.

So, in `/mdx`, under the `Cols[1].tuples[3].children` (that dimension highlighted on the screenshot) you will find these info about rows:

![2018-01-28_214004](https://user-images.githubusercontent.com/4989256/35486433-a1d4ee04-0476-11e8-8edb-dd2d39a12f4c.png)

But under the `/DataSource`, at `rowLevels[3].childLevels` the **structure** is in slightly different format (it joins to a single dimension):

![2018-01-28_214154](https://user-images.githubusercontent.com/4989256/35486456-00dec348-0477-11e8-9b1f-c09719f06f4b.png)

And this difference has no keys/id/whatever to merge it together (relate them somehow). If you think it has, think about label dimensions or dimensions with no data (which are not even present in `/mdx` response): it seems like there's no way to merge them yet.

@eduard93, can MDX2JSON somehow put identifiers on these dimensions to make front end able to figure out how to merge them? Yet, I have seen that MDX2JSON returns some "info" (namely, cells formatting style) right in the `/mdx` tuples response. But it never returned headers info there, I am wondering whether it is even possible. Still, there are a lot of other info to return. Please, let me know what do you think. Thank you!

**Comment by eduard93** [comment]
MDX:
````
{"MDX":"SELECT NON EMPTY {[DateOfSale].[Actual].[YearSold].&[2013],%LABEL([DateOfSale].[Actual].[YearSold].&[2014],\"\",\"\",,\"background:rgb(255, 128, 255);\",\"color:rgb(0, 0, 0);background:rgb(240, 0, 240);font-family:\"\"courier new\"\";font-size:15pt;\"),NONEMPTYCROSSJOIN([DateOfSale].[Actual].[YearSold].&[2015],{[DateOfSale].[Actual].[MonthSold].&[201505],%LABEL([DateOfSale].[Actual].[MonthSold].&[201507],\"\",\"\",,,\"background:rgb(240, 0, 0);\")}),NONEMPTYCROSSJOIN([DateOfSale].[Actual].[YearSold].&[2016],{%LABEL([DateOfSale].[Actual].[MonthSold].&[201601],\"\",\"\",,\"background:rgb(255, 255, 128);\",\"background:rgb(255, 255, 208);\"),[DateOfSale].[Actual].[MonthSold].Members,%LABEL([DateOfSale].[Actual].[MonthSold].&[201608],\"\",\"\",,\"background:rgb(240, 0, 0);\",\"background:rgb(240, 0, 0);\"),%LABEL([DateOfSale].[Actual].[MonthSold].&[201607],\"\",\"\",,\"background:rgb(255, 255, 208);\",\"background:rgb(240, 240, 0);\")})} ON 0,NON EMPTY HEAD({%LABEL([Product].[P1].[Product Name].&[SKU-192],\"\",\"\",,\"color:rgb(0, 0, 0);background:rgb(255, 128, 128);\",\"font-weight: bold;background:rgb(240, 0, 0);\"),%LABEL([Product].[P1].[Product Name].&[SKU-709],\"\",\"\",,\"color:rgb(0, 0, 0);background:rgb(255, 255, 128);\",\"color:rgb(0, 0, 0);background:rgb(240, 240, 0);font-family:verdana;\"),%LABEL([Product].[P1].[Product Name].&[SKU-451],\"\",\"\",,\"font-weight: bold;font-style: italic;color:rgb(0, 0, 0);background:rgb(128, 255, 128);\",\"color:rgb(0, 0, 0);background:rgb(0, 240, 0);\"),%LABEL(NONEMPTYCROSSJOIN([Product].[P1].[Product Category].&[Pasta],{[Product].[P1].[Product Name].&[SKU-287],%LABEL([Product].[P1].[Product Name].&[SKU-296],\"\",\"\",,\"font-weight: bold;color:rgb(0, 0, 240);background:rgb(176, 255, 255);\",\"color:rgb(0, 0, 0);background:rgb(0, 240, 240);\"),[Product].[P1].[Product Name].&[SKU-928],%LABEL([Product].[P1].[Product Name].Members,\"\",\"\",,,\"background:rgb(255, 128, 128);\"),%LABEL([Product].[P1].[Product Name].Members,\"\",\"\",,,\"background:rgb(255, 255, 128);\"),%LABEL([Product].[P1].[Product Name].Members,\"\",\"\",,,\"background:rgb(128, 255, 128);\")}),\"\",\"\",,,\"font-weight: bold;background:rgb(128, 128, 255);\")},2000) ON 1 FROM [HOLEFOODS]"}
````

Added `debugStyle` property to each cell (on a test server). @ZitRos, does it contain relevant information (note that it's not equal to style property)?

![image](https://user-images.githubusercontent.com/5127457/35506595-db2375b6-04fa-11e8-938e-82d943813d3d.png)

**Comment by nikitaeverywhere** [comment]
@eduard93, thank you! It seems it does. Styles of rows/cols/headers are pretty much the only thing I tried to extract from the pivot definition's `rowLevels`/`colLevels` for now. But there may be more properties we will need in the future.

Then, can we name that property as `"headerStyle"` (former `debugStyle`) and `"cellStyle"` (former `style`)? Do not remove "style" prop just yet. Property `"info"`, which copies `"style"` pretty much, can be dropped, LPT don't use it (and DSW either).

**Comment by eduard93** [comment]
so:
- `style`→ `style`
- `debugStyle` → `headerStyle`
- `info`→ `cellStyle`

@gnibeda? Ideas? Comments?

**Comment by nikitaeverywhere** [comment]
Yep, just do not remove `style` prop itself, as at least LPT [depend on it](https://github.com/intersystems-ru/LightPivotTable/blob/dea08fb579c161dd24f1f6f247cf91c9264eb123/source/js/DataController.js#L168). Please, temporary make `style` and `cellStyle` props which are identical. (`cellStyle` just preserves better naming convention)

I'll migrate to `cellStyle` soon.

**Comment by nikitaeverywhere** [comment]
@eduard93, we discussed this with Anton @gnibeda,

`style` → `style`
`info` → `info` (turns out our projects depend on this two, we'll refactor them soon)
`debugStyle` → `headerStyle`
+`cellStyle` (a copy of `style` to preserve better naming convention, as it means exactly cells styling)

So, it's just about adding `headerStyle` and `cellStyle` props.

**Comment by eduard93** [comment]
Done, see https://github.com/intersystems-ru/Cache-MDX2JSON/issues/67
